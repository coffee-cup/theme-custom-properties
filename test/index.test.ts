import {
  generateCustomProperties,
  transformThemeToCustomProperties,
} from "../src";

const darkTheme = {
  hello: "world",
  colors: {
    primary: "white",
    background: "black",
    grey: {
      100: "#fff",
      500: "black",
    },
  },
  shadows: {
    box: "2px 2px 2px black",
  },
};

const lightTheme = {
  ...darkTheme,
  colors: {
    ...darkTheme.colors,
    primary: "black",
    background: "white",
  },
};

describe("custom properties", () => {
  it("generates custom properties for top level values", () => {
    const theme = { hello: "world", foo: 1 };
    expect(generateCustomProperties(theme)).toEqual({
      customProperties: {
        "--hello": "world",
      },
      transformedTheme: {
        hello: "var(--hello)",
      },
    });
  });

  it("generates custom properties for nested theme", () => {
    expect(generateCustomProperties(darkTheme)).toEqual({
      customProperties: {
        "--hello": "world",
        "--colors-primary": "white",
        "--colors-background": "black",
        "--colors-grey-100": "#fff",
        "--colors-grey-500": "black",
        "--shadows-box": "2px 2px 2px black",
      },
      transformedTheme: {
        hello: "var(--hello)",
        colors: {
          background: "var(--colors-background)",
          grey: {
            "100": "var(--colors-grey-100)",
            "500": "var(--colors-grey-500)",
          },
          primary: "var(--colors-primary)",
        },
        shadows: {
          box: "var(--shadows-box)",
        },
      },
    });
  });
});

describe("body css", () => {
  it("generates CSS with default options", () => {
    expect(
      transformThemeToCustomProperties({
        dark: darkTheme,
        light: lightTheme,
      }),
    ).toMatchSnapshot();
  });

  it("generates CSS with dark default theme", () => {
    expect(
      transformThemeToCustomProperties(
        {
          dark: darkTheme,
          light: lightTheme,
        },
        {
          defaultTheme: "dark",
        },
      ),
    ).toMatchSnapshot();
  });

  it("generates CSS with class attribute", () => {
    expect(
      transformThemeToCustomProperties(
        {
          dark: darkTheme,
          light: lightTheme,
        },
        {
          attribute: "class",
        },
      ),
    ).toMatchSnapshot();
  });
});
