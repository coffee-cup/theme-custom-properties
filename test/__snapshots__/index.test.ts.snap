// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`body css generates CSS with class attribute 1`] = `
Object {
  "bodyCSS": ":root {
--hello: world;
--colors-primary: black;
--colors-background: white;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}

.dark {
--hello: world;
--colors-primary: white;
--colors-background: black;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}

.light {
--hello: world;
--colors-primary: black;
--colors-background: white;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}",
  "transformedThemes": Object {
    "dark": Object {
      "colors": Object {
        "background": "var(--colors-background)",
        "grey": Object {
          "100": "var(--colors-grey-100)",
          "500": "var(--colors-grey-500)",
        },
        "primary": "var(--colors-primary)",
      },
      "hello": "var(--hello)",
      "shadows": Object {
        "box": "var(--shadows-box)",
      },
    },
    "light": Object {
      "colors": Object {
        "background": "var(--colors-background)",
        "grey": Object {
          "100": "var(--colors-grey-100)",
          "500": "var(--colors-grey-500)",
        },
        "primary": "var(--colors-primary)",
      },
      "hello": "var(--hello)",
      "shadows": Object {
        "box": "var(--shadows-box)",
      },
    },
  },
}
`;

exports[`body css generates CSS with dark default theme 1`] = `
Object {
  "bodyCSS": ":root {
--hello: world;
--colors-primary: white;
--colors-background: black;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}

[data-theme=\\"dark\\"] {
--hello: world;
--colors-primary: white;
--colors-background: black;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}

[data-theme=\\"light\\"] {
--hello: world;
--colors-primary: black;
--colors-background: white;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}",
  "transformedThemes": Object {
    "dark": Object {
      "colors": Object {
        "background": "var(--colors-background)",
        "grey": Object {
          "100": "var(--colors-grey-100)",
          "500": "var(--colors-grey-500)",
        },
        "primary": "var(--colors-primary)",
      },
      "hello": "var(--hello)",
      "shadows": Object {
        "box": "var(--shadows-box)",
      },
    },
    "light": Object {
      "colors": Object {
        "background": "var(--colors-background)",
        "grey": Object {
          "100": "var(--colors-grey-100)",
          "500": "var(--colors-grey-500)",
        },
        "primary": "var(--colors-primary)",
      },
      "hello": "var(--hello)",
      "shadows": Object {
        "box": "var(--shadows-box)",
      },
    },
  },
}
`;

exports[`body css generates CSS with default options 1`] = `
Object {
  "bodyCSS": ":root {
--hello: world;
--colors-primary: black;
--colors-background: white;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}

[data-theme=\\"dark\\"] {
--hello: world;
--colors-primary: white;
--colors-background: black;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}

[data-theme=\\"light\\"] {
--hello: world;
--colors-primary: black;
--colors-background: white;
--colors-grey-100: #fff;
--colors-grey-500: black;
--shadows-box: 2px 2px 2px black;
}",
  "transformedThemes": Object {
    "dark": Object {
      "colors": Object {
        "background": "var(--colors-background)",
        "grey": Object {
          "100": "var(--colors-grey-100)",
          "500": "var(--colors-grey-500)",
        },
        "primary": "var(--colors-primary)",
      },
      "hello": "var(--hello)",
      "shadows": Object {
        "box": "var(--shadows-box)",
      },
    },
    "light": Object {
      "colors": Object {
        "background": "var(--colors-background)",
        "grey": Object {
          "100": "var(--colors-grey-100)",
          "500": "var(--colors-grey-500)",
        },
        "primary": "var(--colors-primary)",
      },
      "hello": "var(--hello)",
      "shadows": Object {
        "box": "var(--shadows-box)",
      },
    },
  },
}
`;
